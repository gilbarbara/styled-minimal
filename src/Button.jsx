import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { withComponent } from './utils/helpers';
import { buttonType, sizeTypeFull, variantType } from './utils/propTypes';
import { ButtonStyles } from './utils/system';

import { StyledBadge } from './Badge';

export const StyledButton = styled.button`
  ${ButtonStyles.base};
  
  &:disabled {
    pointer-events: none;
  }
  
  &:hover {
    transition: transform 0.2s;
    transform: scale(1.05);
  }
  
  &:focus {  
      ${ButtonStyles.outlineColor};
  }
  
  ${StyledBadge} {
    margin-left: 5px;
  }
`;

StyledButton.propTypes = {
  /** boolean indicating whether the button should render with an animation */
  animate: PropTypes.bool,
  as: PropTypes.string,
  /** match its parent width */
  block: PropTypes.bool,
  children: PropTypes.node.isRequired,
  /** boolean indicating whether the button should render as disabled */
  disabled: PropTypes.bool,
  /** callback on the click event */
  onClick: PropTypes.func,
  outline: PropTypes.bool,
  size: sizeTypeFull,
  type: buttonType,
  variant: variantType,
};

StyledButton.defaultProps = {
  animate: false,
  block: false,
  disabled: false,
  outline: false,
  size: 'md',
  type: 'button',
  variant: 'primary',
};

export default withComponent(StyledButton, 'Button');
