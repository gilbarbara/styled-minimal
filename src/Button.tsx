import * as React from 'react';
import styled, { css } from 'styled-components';

import { getColor, getTheme, px } from './utils/helpers';
import { baseStyles, sx, system, variantStyles } from './utils/system';

import { StyledBadge } from './Badge';

import { ButtonProps } from './types';

export const StyledButton = styled('button')<ButtonProps>(
  props => {
    const { block, busy, invert, size = 'md' } = props;
    const { borderRadius, disabledOpacity, fontSize, lineHeight, loader, padding } = getTheme(
      props,
      'button',
    );

    return css`
      ${baseStyles};
      ${variantStyles};
      ${busy ? loader(invert ? '#ccc' : '#fff') : ''};
      align-items: center;
      border-radius: ${px(borderRadius[size])};
      box-shadow: none;
      cursor: pointer;
      display: inline-flex;
      font-family: inherit;
      font-size: ${px(fontSize[size])};
      line-height: ${lineHeight};
      padding: ${px(padding[size][0])} ${px(padding[size][1])};
      text-decoration: none;
      width: ${block ? '100%' : 'auto'};

      &:disabled {
        opacity: ${disabledOpacity};
        pointer-events: none;
      }

      &:focus {
        outline-color: ${getColor};
      }

      ${StyledBadge} {
        margin-left: 5px;
      }
    `;
  },
  sx,
  system,
);

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => (
  <StyledButton ref={ref} {...props} />
));

Button.defaultProps = {
  block: false,
  busy: false,
  dark: false,
  disabled: false,
  invert: false,
  size: 'md',
  type: 'button',
  variant: 'primary',
};
