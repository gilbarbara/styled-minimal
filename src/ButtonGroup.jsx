import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { sizesAllPropTypes, variantPropTypes } from './utils/system';

import Box from './Box';

export const StyledButtonGroup = styled(Box)`
  display: inline-flex;

  > button {
    + button {
      margin-left: -1px;
    }

    &:first-child {
      border-bottom-right-radius: 0;
      border-top-right-radius: 0;
    }

    &:last-child {
      border-bottom-left-radius: 0;
      border-top-left-radius: 0;
    }

    &:not(:first-child):not(:last-child) {
      border-radius: 0;
    }
  }
`;

const ButtonGroup = ({ children, size, variant, ...props }) => {
  const buttonProps = {};

  if (typeof size !== 'undefined') {
    buttonProps.size = size;
  }

  if (typeof variant !== 'undefined') {
    buttonProps.variant = variant;
  }

  return (
    <StyledButtonGroup {...props}>
      {React.Children.map(children, child => React.cloneElement(child, { ...buttonProps }))}
    </StyledButtonGroup>
  );
};

ButtonGroup.displayName = 'ButtonGroup';

ButtonGroup.propTypes = {
  children: PropTypes.node.isRequired,
  size: sizesAllPropTypes,
  variant: variantPropTypes,
  ...Box.basePropTypes,
};

export default ButtonGroup;
