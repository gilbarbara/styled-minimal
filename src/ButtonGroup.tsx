import * as React from 'react';
import styled, { css } from 'styled-components';

import { baseStyles, sx, system } from './utils/system';

import { ButtonGroupProps } from './types';

export const StyledButtonGroup = styled('div')<Partial<ButtonGroupProps>>(
  () => css`
    ${baseStyles};
    display: inline-flex;

    > button {
      + button {
        margin-left: -1px;
      }

      &:first-child {
        border-bottom-right-radius: 0;
        border-top-right-radius: 0;
      }

      &:last-child {
        border-bottom-left-radius: 0;
        border-top-left-radius: 0;
      }

      &:not(:first-child):not(:last-child) {
        border-radius: 0;
      }
    }
  `,
  sx,
  system,
);

export const ButtonGroup = React.forwardRef<HTMLDivElement, ButtonGroupProps>(
  ({ children, size, variant, ...rest }: ButtonGroupProps, ref) => {
    const buttonProps = {
      size,
      variant,
    };

    return (
      <StyledButtonGroup ref={ref} {...rest}>
        {React.Children.map(children, child =>
          React.cloneElement(child as React.ReactElement, { ...buttonProps }),
        )}
      </StyledButtonGroup>
    );
  },
);

ButtonGroup.defaultProps = {
  size: 'md',
  variant: 'primary',
};
