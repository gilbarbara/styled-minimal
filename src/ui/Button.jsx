import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { getProp } from '../utils/helpers';
import { colors } from '../utils/styles';

const Wrapper = styled.button.attrs({
  type: 'button',
})`
  align-items: center;
  border-radius: ${getProp('buttonRadius', 'size', 'md')};
  background-color: ${getProp('palette', 'variant', 'primary')};
  border: 0;
  color: ${({ variant }) => (variant === 'default' ? colors.black : colors.white)};
  display: flex;
  font-size: ${getProp('fontSizes', 'size', 'md')};
  font-weight: 700;
  padding: ${getProp('buttonPadding', 'size', 'md')};
  text-transform: uppercase;
  ${({ loading }) => (loading ? getProp('buttonLoader') : '')}
  
  [class^=i-] {
    margin-right: 1rem;
  }
`;

const Button = ({ children, ...props }) => (
  <Wrapper {...props}>{children}</Wrapper>
);

Button.propTypes = {
  children: PropTypes.node.isRequired,
  loading: PropTypes.bool,
  size: PropTypes.string,
  variant: PropTypes.string,
};

Button.defaultProps = {
  size: 'md',
  variant: 'primary',
};

export default Button;
