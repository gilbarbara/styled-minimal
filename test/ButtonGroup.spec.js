import React from 'react';
import { mount } from 'enzyme';

import Button from '../src/Button';
import ButtonGroup from '../src/ButtonGroup';

describe('ButtonGroup', () => {
  const wrapper = mount(
    <ButtonGroup
      size="sm"
      variant="indigo"
    >
      <Button outline>
        First
      </Button>
      <Button>
        Second
      </Button>
      <Button outline>
        Third
      </Button>
      <Button outline>
        Forth
      </Button>
    </ButtonGroup>
  );

  it('should render properly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
