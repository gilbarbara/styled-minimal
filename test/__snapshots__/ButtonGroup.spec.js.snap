// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup should render properly 1`] = `
.c1 {
  box-sizing: border-box;
  background-color: #fff;
  border: 1px solid #3F51B5;
  color: #3F51B5;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0px;
  box-shadow: none;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: inherit;
  font-size: 14px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1;
  padding: 7px 14px;
  -webkit-text-decoration: none;
  text-decoration: none;
  width: auto;
}

.c1:disabled {
  pointer-events: none;
}

.c1:hover {
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
}

.c1:focus {
  outline-color: #3F51B5;
}

.c1 .c3 {
  margin-left: 5px;
}

.c2 {
  box-sizing: border-box;
  background-color: #3F51B5;
  border: 1px solid #3F51B5;
  color: #fff;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0px;
  box-shadow: none;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: inherit;
  font-size: 14px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1;
  padding: 7px 14px;
  -webkit-text-decoration: none;
  text-decoration: none;
  width: auto;
}

.c2:disabled {
  pointer-events: none;
}

.c2:hover {
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
}

.c2:focus {
  outline-color: #3F51B5;
}

.c2 .c3 {
  margin-left: 5px;
}

.c0 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c0 > button + button {
  margin-left: -1px;
}

.c0 > button:first-child {
  border-bottom-right-radius: 0;
  border-top-right-radius: 0;
}

.c0 > button:last-child {
  border-bottom-left-radius: 0;
  border-top-left-radius: 0;
}

.c0 > button:not(:first-child):not(:last-child) {
  border-radius: 0;
}

<StyledButtonGroup>
  <div
    className="c0"
  >
    <Button
      animate={false}
      block={false}
      disabled={false}
      key=".0"
      outline={true}
      size="sm"
      type="button"
      variant="indigo"
    >
      <div
        className="c1"
        disabled={false}
        size="sm"
        type="button"
      >
        First
      </div>
    </Button>
    <Button
      animate={false}
      block={false}
      disabled={false}
      key=".1"
      outline={false}
      size="sm"
      type="button"
      variant="indigo"
    >
      <div
        className="c2"
        disabled={false}
        size="sm"
        type="button"
      >
        Second
      </div>
    </Button>
    <Button
      animate={false}
      block={false}
      disabled={false}
      key=".2"
      outline={true}
      size="sm"
      type="button"
      variant="indigo"
    >
      <div
        className="c1"
        disabled={false}
        size="sm"
        type="button"
      >
        Third
      </div>
    </Button>
    <Button
      animate={false}
      block={false}
      disabled={false}
      key=".3"
      outline={true}
      size="sm"
      type="button"
      variant="indigo"
    >
      <div
        className="c1"
        disabled={false}
        size="sm"
        type="button"
      >
        Forth
      </div>
    </Button>
  </div>
</StyledButtonGroup>
`;
